docs/agent_config.md
# Agent Configuration

## Global

```yaml
interval: 5m0s # duration each process to fetch certificates. default: 5m
http:
    listen: 0.0.0.0:8080 # http server listen address. default: 0.0.0.0:8080
    metrics_enable: false # enable metrics on path `/metrics`. default: false
    tls:
        enable: false
        listen: 0.0.0.0:443 # https server listen address.
        cert_path: "/ssl/certificate.crt" # mandatory only when enable is true
        key_path: "/ssl/private.key" # mandatory only when enable is true
manager:
    address: 127.0.0.1:8080 # server address
    token: tokenJwt # JWT token used to authenticate on server
```

## State

State is used to save all certificates.

### Filesystem (default)

```yaml
state:
    type: fs
    config:
        path: /var/lib/lets-go-tls/state.json # mandatory
```

## Requesters

A Requester defines the method by which certificate requests are retrieved.

### Static

```yaml
requesters:
    - id: static
      type: static
      config:
          domains: # mandatory
            - - example.com
            - - foo.com
              - bar.com
```

### Traefik (V2 & V3)

```yaml
requesters:
    - id: traefik
      type: traefikV2 # value can be traefikV2 or traefikV3
      config:
          addresses: # mandatory
              - http://127.0.0.1
              - http://127.0.0.1:8080
              - https://127.0.0.1/api
```

### Nginx

Nginx config is parsed and find all domains in `server_name` when directives `ssl_certificate` and `ssl_certificate_key` are defined in `server` block.

```yaml
requesters:
      - id: nginx
        type: nginx
        config:
          nginx_cfg_path: /etc/nginx/nginx.conf
```

Note: Vhost with server name who contains invalid characters example: `server_name (foo|bar).example.com;` or `server_name *.example.com;` will be ignored.

## Storages

A storage defines where certificate must be stored.

### Filesystem

```yaml
storages:
    - id: fs
      type: fs
      config:
          path: /var/lib/lets-go-tls/ssl # mandatory
          prefix_filename: "ssl."
          owner: "root"
          group: "root"
          add_pem: false # when true, add pem file
          only_matched_domains: false # when true, only store certificate specified in specific_domains
          specific_domains:
            - identifier: custom # mandatory
              path: "custom" # optional, can be absolut or relative
              domains: # mandatory
                - bar.com
          post_hook:
            cmd: "bash -c 'nginx -t && systemctl reload nginx'" # mandatory, run command when certificates have changed
            timeout: 1m0s
```

tree structure example:

```
/var/lib/lets-go-tls
└── ssl/
    ├── ssl.example.com-0.key
    ├── ssl.example.com-0.crt
    ├── ssl.foo.com-0.key
    └── ssl.foo.com-0.crt
    ├── ssl.custom.key
    └── ssl.custom.crt
```

### Traefik (V2 & V3)

Traefik configurations:
* [static configuration](./examples/traefik/config.yml)
* [dynamic configuration](./examples/traefik/router.yml)

```yaml
storages:
    - id: traefik
      type: traefikV2 # value can be traefikV2 or traefikV3
      config:
          path: /var/lib/traefik/ssl # mandatory
          prefix_filename: "ssl."
          owner: "root"
          group: "root"
```

tree structure example:
```
/var/lib/traefik
└── ssl/
    ├── ssl.example.com-0.yml
    └── ssl.foo.com-0.yml
```

### Haproxy

```yaml
storages:
    - id: haproxy
      type: haproxy
      config:
          path: /etc/haproxy/ssl # mandatory
          crt_list_path: /etc/haproxy/crt-list.txt # optional, default value: {{ path }}/crt-list.txt
          prefix_filename: "ssl."
          owner: "root"
          group: "root"
          add_pem: true # value forced to true
          only_matched_domains: false # when true, only store certificate specified in specific_domains
          specific_domains:
            - identifier: custom # mandatory
              path: "custom" # optional, can be absolut or relative
              domains: # mandatory
                - bar.com
          post_hook:
            cmd: "bash -c 'systemctl reload haproxy'" # mandatory, run command when certificates have changed
            timeout: 1m0s
```

tree structure example:

```
/etc/haproxy
├── crt-list.txt
└── ssl/
    ├── ssl.example.com-0.key
    ├── ssl.example.com-0.crt
    ├── ssl.example.com-0.pem
    ├── ssl.foo.com-0.key
    ├── ssl.foo.com-0.crt
    ├── ssl.foo.com-0.pem
    ├── ssl.custom.key
    ├── ssl.custom.crt
    └── ssl.custom.pem
```

### Nginx

Nginx config is parsed and write certificates and keys in `ssl_certificate` and `ssl_certificate_key` path, based on `server_name` domains.

```yaml
storages:
      - id: nginx
        type: nginx
        config:
          nginx_cfg_path: /etc/nginx/nginx.conf # mandatory
          group: root
          owner: root
          post_hook:
            cmd: "bash -c 'systemctl reload nginx'" # mandatory, run command when certificates have changed
            timeout: 1m0s
```

Note: Vhost with server name who contains invalid characters example: `server_name (foo|bar).example.com;` or `server_name *.example.com;` will be ignored.


docs/server_config.md
# Server Configuration

## Global

```yaml
interval: 5m0s # duration each process to fetch requesters and obtain certificate. default: 5m
lock_duration: 25m0s # max duration to lock process to obtain or renew certificate to prevent concurrency. default: 25m
unused_retention: 336h0m0s # time to keep in store unused certificate. default: 14 days
http:
    listen: 0.0.0.0:8080 # http server listen address. default: 0.0.0.0:8080
    metrics_enable: false # enable metrics on path `/metrics`. default: false
    tls:
        enable: false
        listen: 0.0.0.0:443 # https server listen address.
        cert_path: "/ssl/certificate.crt" # mandatory only when enable is true
        key_path: "/ssl/private.key" # mandatory only when enable is true
jwt:
    key: superSecret # secret to sign JWT token
    method: HS256 # method used to sign JWT token. default: HS256
acme:
    ca_server: https://acme-v02.api.letsencrypt.org/directory # CA server address. default: https://acme-v02.api.letsencrypt.org/directory
    email: acme@example.com # email used for ACME registration
    renew_period: 240h0m0s # period before the end of a certificate. default: 10 days
    delay_failed: 24h0m0s # delay when a certificate reach max fail attempt to obtain or renew. default: 24h 
    max_attempt: 3 # max attempt when a certificate fail to obtain or renew. default: 3
```

## State

State is used to save ACME account and all certificates.

### Filesystem (default)

```yaml
state:
    type: fs
    config:
        path: /var/lib/lets-go-tls/state.json # mandatory
```

## Requesters

A Requester defines the method by which certificate requests are retrieved.

### Static

```yaml
requesters:
    - id: static
      type: static
      config:
          domains: # mandatory
            - - example.com
            - - foo.com
              - bar.com
```

### Agent

```yaml
requesters:
    - id: agents
      type: agent
      config:
          addresses: # mandatory
              - 127.0.0.1:8080
```

## Cache

Cache is used to prevent concurrency process.

### Memory (default)

```yaml
cache:
    type: memory
```

### Redis

```yaml
cache:
    type: redis
    config:
        address: 127.0.0.1:6379 # mandatory
        db: my_db
        username: user
        password: password
```

### Redis cluster

```yaml
cache:
    type: redis-cluster
    config:
        address: # mandatory
          - 127.0.0.1:6379 
          - 127.0.0.2:6379
        username: user
        password: password
```

## ACME resolvers

By default, http ACME challenge is used.
The server must receive all URI `/.well-known/acme-challenge` for ACME Challenge. 

### DNS Challenges

The key `filters` define domains who must use specific resolver.

Examples:
* Whole domain:
    * filters: example.com
    * match: example and all subdomain
* Only subdomain:
    * filters: sub.example.com
    * match: sub.example.com and *.sub.example.com but don't match foo.example.com or example.com

```yaml
acme:
  resolvers:
      gandiv5:
          type: gandiv5
          config:
              api_key: ""
              http_timeout: 3m0s
              polling_interval: 2s
              propagation_timeout: 1m0s
          filters:
              - foo.com
      httpreq:
          type: httpreq
          config:
              endpoint: ""
              http_timeout: 3m0s
              mode: ""
              password: ""
              polling_interval: 2s
              propagation_timeout: 1m0s
              username: ""
          filters:
              - foo.com
      ovh:
          type: ovh
          config:
              access_token: ""
              application_key: ""
              application_secret: ""
              client_id: ""
              client_secret: ""
              consumer_key: ""
              endpoint: ovh-eu
              http_timeout: 3m0s
              polling_interval: 2s
              propagation_timeout: 1m0s
          filters:
              - foo.com
```

If you need other resolver you can open an issue or a pull request.


apps/server/config/config.go
package config

import (
	"github.com/alexandreh2ag/lets-go-tls/config"
	"github.com/go-acme/lego/v4/lego"
	"time"
)

type Config struct {
	Requesters              []config.RequesterConfig `mapstructure:"requesters" validate:"required,unique=Id,dive"`
	Acme                    AcmeConfig               `mapstructure:"acme" validate:"required"`
	State                   config.StateConfig       `mapstructure:"state" validate:"required"`
	Cache                   CacheConfig              `mapstructure:"cache" validate:"required"`
	HTTP                    config.HTTPConfig        `mapstructure:"http" validate:"required"`
	JWT                     JWTConfig                `mapstructure:"jwt" validate:"required"`
	Interval                time.Duration            `mapstructure:"interval" validate:"required"`
	LockDuration            time.Duration            `mapstructure:"lock_duration" validate:"required"`
	UnusedRetentionDuration time.Duration            `mapstructure:"unused_retention" validate:"required"`
}

type AcmeConfig struct {
	CAServer    string                    `mapstructure:"ca_server" validate:"required"`
	Email       string                    `mapstructure:"email" validate:"required,email"`
	Resolvers   map[string]ResolverConfig `mapstructure:"resolvers,omitempty"`
	RenewPeriod time.Duration             `mapstructure:"renew_period" validate:"required"`
	MaxAttempt  int                       `mapstructure:"max_attempt" validate:"required,min=1"`
	DelayFailed time.Duration             `mapstructure:"delay_failed" validate:"required"`
}

type CacheConfig struct {
	Type   string                 `mapstructure:"type" validate:"required,excludesall=!@#$ "`
	Config map[string]interface{} `mapstructure:"config,omitempty"`
}

type ResolverConfig struct {
	Type    string                 `mapstructure:"type" validate:"required,excludesall=!@#$ "`
	Config  map[string]interface{} `mapstructure:"config"`
	Filters []string               `mapstructure:"filters" validate:"required,min=1"`
}

type JWTConfig struct {
	Key    string `mapstructure:"key" validate:"required,min=1"`
	Method string `mapstructure:"method" validate:"required"`
}

func NewConfig() Config {
	return Config{}
}

func DefaultConfig() Config {
	cfg := NewConfig()
	cfg.Interval = time.Minute * 5
	cfg.LockDuration = time.Minute * 25
	cfg.UnusedRetentionDuration = time.Hour * 24 * 14
	cfg.HTTP = config.HTTPConfig{Listen: "0.0.0.0:8080"}
	cfg.Cache = CacheConfig{Type: "memory"}
	cfg.Acme = AcmeConfig{
		CAServer:    lego.LEDirectoryProduction,
		Resolvers:   map[string]ResolverConfig{},
		RenewPeriod: time.Hour * 24 * 10,
		MaxAttempt:  3,
		DelayFailed: time.Hour * 24,
	}
	cfg.JWT = JWTConfig{Method: "HS256"}
	return cfg
}
